### PR TITLE
Add `CategoricalValue(x, source)` and disallow mixed `isless` and `<`

### LINKED ISSUE #319: Generic test for allowing to use isless
Consider the following example:
```
julia> x = [1, 2]
2-element Array{Int64,1}:
 1
 2

julia> y = categorical([1])
1-element CategoricalArray{Int64,1,UInt32}:
 1

julia> isless(x[1], y[1])
false

julia> isless(x[2], y[1])
ERROR: KeyError: key 2 not found
```
This behaviour is very problematic I think. The reason is that given two containers `x` and `y` there is no way to check if you can safely use `isless` to compare their elements. I could run a strict check if `eltype` of `x` and `y` are the same, but it is not desirable very often (e.g. numbers can be compared even if their type is different).

### PR DESCRIPTION
**Pull request by nalimilan**
`isless` and `<` between `CategoricalValue` and other types can breaks transitivity. Require converting the value to a `CategoricalValue`, attaching it to a pool, using `CategoricalValue(x, source::Union{CategoricalValue, CategoricalArray)`.
Deprecate `CategoricalValue(i::Integer, pool::CategoricalPool)` in favor of `pool[i]` or `CategoricalValue(pool, i)`, as the order is more logical and it avoids a potential ambiguity if `source::CategoricalPool` is allowed in the future.

Fixes #319.

### CODE REVIEW

File: src/missingarray.jl [hunk 1]
```diff
@@ -34,10 +34,10 @@ in(x::Missing, y::CategoricalArray{>:Missing}) = !all(v -> v > 0, y.refs)
 
 function Missings.replace(a::CategoricalArray{S, N, R, V, C}, replacement::V) where {S, N, R, V, C}
     pool = copy(a.pool)
-    v = C(get!(pool, replacement), pool)
+    v = C(pool, get!(pool, replacement))
     Missings.replace(a, v)
 end
 
 function collect(r::Missings.EachReplaceMissing{<:CategoricalArray{S, N, R, C}}) where {S, N, R, C}
```
Comment (bkamins) [inline src/missingarray.jl:41]: it is a bit confusing that you use `C` to mean what is named `V` above. Maybe make it consistent?

File: src/typedefs.jl [hunk 2]
```diff
@@ -82,8 +82,8 @@ the order of the pool's [`levels`](@ref DataAPI.levels) is used rather than the
 ordering of values of type `T`.
 """
 struct CategoricalValue{T <: SupportedTypes, R <: Integer}
-    level::R
     pool::CategoricalPool{T, R, CategoricalValue{T, R}}
+    ref::R
```
Comment (bkamins) [inline src/typedefs.jl:86]: smart - I was thinking how you have changed it - and it is just super simple :).

File: src/value.jl [hunk 3]
```diff
@@ -140,8 +153,11 @@ function Base.isless(x::CategoricalValue, y::CategoricalValue)
     end
 end
 
-Base.isless(x::CategoricalValue, y::SupportedTypes) = levelcode(x) < levelcode(x.pool[get(x.pool, y)])
-Base.isless(y::SupportedTypes, x::CategoricalValue) = levelcode(x.pool[get(x.pool, y)]) < levelcode(x)
+Base.isless(x::CategoricalValue, y::SupportedTypes) =
```
Comment (bkamins) [inline src/value.jl:156]: so you have decided to go forward fast and be breaking here? (I am OK with this, but usually you wanted one round of releases with deprecations :smile: )

**Review by bkamins — APPROVED** [review]
Looks good except for small comments

File: src/value.jl (hunk 3 above)
Comment (nalimilan) [reply src/value.jl:156]: Yes maybe I can backport a deprecation to 0.9. Though the error message is basically the same as a deprecation so I wonder whether it's worth it.

File: src/missingarray.jl (hunk 1 above)
Comment (nalimilan) [reply src/missingarray.jl:41]: Indeed. Better always use the same names as in the type definition.

File: src/typedefs.jl (hunk 2 above)
Comment (nalimilan) [reply src/typedefs.jl:86]: Actually I wanted to change their order for a long time, but there's never been a strong enough reason to do it. :-)

File: src/value.jl (hunk 3 above)
Comment (bkamins) [reply src/value.jl:156]: I am OK with error as commented.

**Pull request merged**